pkg/model/components/crio.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package components
+
+import (
+	"fmt"
+
+	"github.com/blang/semver"
+	"github.com/pelletier/go-toml"
+	"k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/loader"
+)
+
+// CrioOptionsBuilder adds options for crio to the model
+type CrioOptionsBuilder struct {
+	*OptionsContext
+}
+
+var _ loader.OptionsBuilder = &CrioOptionsBuilder{}
+
+// BuildOptions is responsible for filling in the default setting for crio daemon
+func (c *CrioOptionsBuilder) BuildOptions(o interface{}) error {
+	clusterSpec := o.(*kops.ClusterSpec)
+
+	if clusterSpec.Crio == nil {
+		clusterSpec.Crio = &kops.CrioConfig{}
+	}
+
+	crio := clusterSpec.Crio
+
+	if clusterSpec.ContainerRuntime == "crio" {
+		if c.IsKubernetesLT("1.21") {
+			return fmt.Errorf("kOps does not support crio with k8s version less than 1.21")
+		}
+
+		// Unless user specifies version, use the cri-o version matching the cluster's kubernetes version
+		if crio.Version == nil {
+			crioVersion, err := c.getCrioVersion(clusterSpec.KubernetesVersion)
+			if err != nil {
+				return err
+			}
+
+			crio.Version = crioVersion
+		}
+
+		if fi.StringValue(crio.ConfigOverride) == "" {
+			crio.ConfigOverride = c.buildCrioDefaultConfig()
+		}
+
+		if fi.StringValue(crio.ContainerPolicyOverride) == "" {
+			crio.ContainerPolicyOverride = c.buildCrioDefaultContainerPolicy()
+		}
+
+		if fi.StringValue(crio.ContainerRegistriesOverride) == "" {
+			crio.ContainerRegistriesOverride = c.buildCrioDefaultContainerRegistriesPolicy()
+		}
+
+		if crio.LogLevel == nil {
+			crio.LogLevel = fi.String("info")
+		}
+
+	} else {
+		crio.SkipInstall = true
```

Comment:
Would be easier to follow if you inverted the sense of the if block and then did an early return here.